src/shell/scripts/omp.ps1

```diff
@@ -120,108 +128,105 @@ New-Module -Name "oh-my-posh-core" -ScriptBlock {
-            $host.UI.RawUI.CursorPosition = $position
-            # cache the tip command
-            $script:ToolTipCommand = $command
-            # we need this workaround to prevent the text after cursor from disappearing when the tooltip is rendered
-            [Microsoft.PowerShell.PSConsoleReadLine]::Insert(' ')
-            [Microsoft.PowerShell.PSConsoleReadLine]::Undo()
+    function Get-TerminalWidth {
+        $terminalWidth = $Host.UI.RawUI.WindowSize.Width
+        # Set a sane default when the value can't be retrieved.
+        if (-not $terminalWidth) {
+            return 0
         }
+        $terminalWidth
     }
 
-    function Set-TransientPrompt {
-        $previousOutputEncoding = [Console]::OutputEncoding
-        $executingCommand = $false
+    if (!$script:ConstrainedLanguageMode) {
+        if ('::TOOLTIPS::' -eq 'true') {
+            Set-PSReadLineKeyHandler -Key Spacebar -BriefDescription 'OhMyPoshSpaceKeyHandler' -ScriptBlock {
+                param([ConsoleKeyInfo]$key)
+                [Microsoft.PowerShell.PSConsoleReadLine]::SelfInsert($key)
+                try {
+                    $command = ''
+                    [Microsoft.PowerShell.PSConsoleReadLine]::GetBufferState([ref]$command, [ref]$null)
+                    # Get the first word of command line as tip.
+                    $command = $command.TrimStart().Split(' ', 2) | Select-Object -First 1
+                    # Ignore an empty/repeated tooltip command.
+                    if (!$command -or ($command -eq $script:TooltipCommand)) {
+                        return
+                    }
+                    $script:TooltipCommand = $command
+                    $position = $Host.UI.RawUI.CursorPosition
+                    $terminalWidth = Get-TerminalWidth
+                    $cleanPSWD = Get-CleanPSWD
+                    $stackCount = global:Get-PoshStackCount
+                    $standardOut = [string]((Start-Utf8Process $script:OMPExecutable @("print", "tooltip", "--status=$script:ErrorCode", "--shell=$script:ShellName", "--pswd=$cleanPSWD", "--execution-time=$script:ExecutionTime", "--stack-count=$stackCount", "--config=$env:POSH_THEME", "--command=$command", "--shell-version=$script:PSVersion", "--column=$($position.X)", "--terminal-width=$terminalWidth", "--no-status=$script:NoExitCode")) -join '')
+                    if (!$standardOut) {
+                        return
+                    }
+
+                    # Save the tooltip prompt to avoid unnecessary CLI calls.
+                    $script:TooltipPrompt = $standardOut.TrimStart()
+                    $paddingLength = $standardOut.Length - $script:TooltipPrompt.Length
+                    $script:TooltipPromptWidth = $terminalWidth - $position.X - $paddingLength
+
+                    # Remove the right-aligned prompt (if any) from the saved prompt.
+                    $script:CurrentPrompt = $script:CurrentPrompt -replace "^([\s\S]+)`e7\s*.+`e8([\s\S]+)?$", '$1$2'
```

Comment:
this feels so hacky. I know it will work but it's a recipe for disaster in my opinion. The previous implementation was easier to understand.

Reply:
This workaround exists because the length of the right prompt padding is calculated in the OMP engine. Or should we make it just like other shells that have native right prompt support? I mean, for PowerShell we can extract the right prompt rendering logic instead of merging it into the primary prompt. Seems less hacky but does require one more CLI call when rendering full prompts. 🤦🏻‍♂️

Reply:
@lewis-yeung we didn't do this before though, it's the prompt caching that introduces this. I wonder if we can do this differently as this isn't code I like seeing. I get the challenge here, but I think we can do better.

Reply:
Reworked and ready for a review now.